File: modules/preflight.py
"""Preflight checks for the pipeline.

Run all the credential and capacity validations BEFORE step 1 starts.
Fails fast on misconfiguration instead of letting a pipeline burn 5
minutes provisioning a CF zone before discovering the SA token expired.

Each check returns:
  {ok: bool, message: str, detail: dict | None}

The aggregate endpoint (/api/preflight/<domain>) returns:
  {ok: bool (all green), checks: {check_name: result}}

Checks are intentionally cheap (single API hit each) and run sequentially
— total budget is ~10s on a healthy network.
"""

from __future__ import annotations

from database import get_setting, get_db


def _result(ok: bool, message: str, detail: dict | None = None) -> dict:
    out = {"ok": ok, "message": message}
    if detail:
        out["detail"] = detail
    return out


# ---------------------------------------------------------------------------
# Individual checks
# ---------------------------------------------------------------------------

def check_cf_pool() -> dict:
    """At least one CF key with capacity left in the pool."""
    conn = get_db()
    try:
        rows = conn.execute(
            """SELECT id, alias, email, domains_used, max_domains
                 FROM cf_keys
                WHERE is_active = 1 AND domains_used < max_domains""",
        ).fetchall()
        total_active = conn.execute(
            "SELECT COUNT(*) FROM cf_keys WHERE is_active = 1"
        ).fetchone()[0]
    finally:
        conn.close()
    if not rows:
        if total_active == 0:
            return _result(False, "No active CF keys in the pool. Add one in Settings.")
        return _result(False,
            f"All {total_active} active CF keys are at capacity. "
            "Add a new key or raise max_domains on an existing one.")
    available = sum((r["max_domains"] - r["domains_used"]) for r in rows)
    return _result(True,
        f"{len(rows)} key(s) with capacity, {available} domain slot(s) free")


def check_do_token() -> dict:
    """Primary DO token reaches /v2/account."""
    import requests
    tok = (get_setting("do_api_token") or "").strip()
    if not tok:
        return _result(False, "DO_API_TOKEN not set in Settings")
    try:
        r = requests.get(
            "https://api.digitalocean.com/v2/account",
            headers={"Authorization": f"Bearer {tok}"},
            timeout=15,
        )
    except Exception as e:
        return _result(False, f"DO API unreachable: {type(e).__name__}: {e}")
    if not r.ok:
        return _result(False, f"DO API rejected token (HTTP {r.status_code})",
                       {"body": r.text[:140]})
    acc = r.json().get("account", {})
    return _result(True,
        f"DO ok ({acc.get('email', '?')}, droplet_limit={acc.get('droplet_limit')})",
        {"email": acc.get("email"),
         "droplet_limit": acc.get("droplet_limit"),
         "status": acc.get("status")})


def check_sa_auth() -> dict:
    """SA token + org id resolve to a real org."""
    import requests
    tok = (get_setting("serveravatar_api_key") or "").strip()
    org = (get_setting("serveravatar_org_id") or "").strip()
    if not tok:
        return _result(False, "SERVERAVATAR_API_KEY not set in Settings")
    if not org:
        return _result(False, "SERVERAVATAR_ORG_ID not set in Settings")
    try:
        r = requests.get(
            f"https://api.serveravatar.com/organizations/{org}",
            headers={"Authorization": tok, "Accept": "application/json"},
            timeout=15,
        )
    except Exception as e:
        return _result(False, f"SA API unreachable: {type(e).__name__}: {e}")
    if not r.ok:
        return _result(False, f"SA API rejected (HTTP {r.status_code})",
                       {"body": r.text[:140]})
    return _result(True, f"SA ok (org={org})")


def check_spaceship_auth(skip_purchase: bool = False) -> dict:
    """Spaceship token works. Skipped if the operator marked skip_purchase
    (they're using a domain registered elsewhere and don't need our
    Spaceship account)."""
    if skip_purchase:
        return _result(True, "Skipped (skip_purchase=True)")
    import requests
    api_key = (get_setting("spaceship_api_key") or "").strip()
    api_secret = (get_setting("spaceship_api_secret") or "").strip()
    if not api_key or not api_secret:
        return _result(False, "SPACESHIP_API_KEY / _SECRET not both set")
    try:
        r = requests.get(
            "https://spaceship.dev/api/v1/domains?take=1",
            headers={"X-API-Key": api_key, "X-API-Secret": api_secret,
                     "Accept": "application/json"},
            timeout=15,
        )
    except Exception as e:
        return _result(False, f"Spaceship unreachable: {type(e).__name__}: {e}")
    if not r.ok:
        return _result(False, f"Spaceship API rejected (HTTP {r.status_code})",
                       {"body": r.text[:140]})
    return _result(True, "Spaceship ok")


def check_llm_key() -> dict:
    """At least one LLM provider key is configured. Doesn't make a billable
    call — just checks the configured provider's key is present and non-empty.
    The full key validation is the existing /api/settings/test-llm-key path."""
    provider = (get_setting("llm_provider") or "anthropic").strip().lower()
    key = ((get_setting(f"llm_api_key_{provider}") or "").strip()
           or (get_setting("llm_api_key") or "").strip())
    if not key:
        return _result(False, f"LLM provider={provider} but no API key configured")
    return _result(True, f"LLM provider={provider} key configured ({len(key)} chars)")


def check_server_capacity() -> dict:
    """At least one ready server below max_sites, or a working DO token to
    provision a new one (we can spin one up if needed)."""
    conn = get_db()
    try:
        rows = conn.execute(
            """SELECT s.id, s.name, s.ip,
                    (SELECT COUNT(*) FROM domains d WHERE d.server_id = s.id) AS sites_count,
                    s.max_sites
                 FROM servers s
                WHERE s.status = 'ready'""",
        ).fetchall()
    finally:
        conn.close()
    rows = [dict(r) for r in rows]
    available = [r for r in rows
                 if (r["sites_count"] or 0) < (r["max_sites"] or 60)]
    if available:
        return _result(True,
            f"{len(available)} ready server(s) with capacity",
            {"servers": [{"name": r["name"], "ip": r["ip"],
                          "sites": r["sites_count"], "max": r["max_sites"]}
                         for r in available]})
    if rows:
        return _result(False,
            f"All {len(rows)} ready server(s) are at max_sites. "
            "Pipeline will try to provision a new droplet.")
    return _result(False,
        "No ready servers. Pipeline will try to provision a new droplet "
        "(requires healthy DO token).")


def check_root_password() -> dict:
    """Server root password is set — needed for SSH password auth fallback
    when provisioning a fresh droplet."""
    pw = (get_setting("server_root_password") or "").strip()
    if not pw:
        return _result(False,
            "server_root_password not set. New droplets won't be reachable "
            "via password SSH for setup. Set in Settings.")
    return _result(True, f"Root password set ({len(pw)} chars)")


# ---------------------------------------------------------------------------
# Aggregate
# ---------------------------------------------------------------------------

def run_all(skip_purchase: bool = False) -> dict:
    """Run every check sequentially. Returns the aggregate dict."""
    checks = {
        "cf_pool":         check_cf_pool(),
        "do_token":        check_do_token(),
        "sa_auth":         check_sa_auth(),
        "spaceship_auth":  check_spaceship_auth(skip_purchase=skip_purchase),
        "llm_key":         check_llm_key(),
        "server_capacity": check_server_capacity(),
        "root_password":   check_root_password(),
    }
    return {
        "ok": all(c["ok"] for c in checks.values()),
        "checks": checks,
    }

File: template_v0/lib/preflight.ts
/**
 * Preflight checks — Node port of modules/preflight.py.
 *
 * Run all credential + capacity validations BEFORE step 1 of the pipeline
 * starts. Each check is a single API hit; the aggregate budget is ~10s on
 * a healthy network. Fails fast on misconfiguration so we don't burn 5min
 * provisioning a CF zone before discovering the SA token expired.
 *
 * Aggregate `runAll()` returns `{ ok, checks: {<name>: <result>} }` —
 * matches the shape the Flask `/api/preflight/<domain>` endpoint returns
 * so the existing dashboard rendering can consume both.
 */

import { existsSync, readdirSync } from "node:fs"
import path from "node:path"
import os from "node:os"
import { all } from "./db"
import { getSetting } from "./repos/settings"

export interface CheckResult {
  ok: boolean
  message: string
  detail?: Record<string, unknown>
}

function ok(message: string, detail?: Record<string, unknown>): CheckResult {
  const r: CheckResult = { ok: true, message }
  if (detail) r.detail = detail
  return r
}
function fail(message: string, detail?: Record<string, unknown>): CheckResult {
  const r: CheckResult = { ok: false, message }
  if (detail) r.detail = detail
  return r
}

// ---------------------------------------------------------------------------
// Individual checks
// ---------------------------------------------------------------------------

export function checkCfPool(): CheckResult {
  const rows = all<{ id: number; alias: string | null; email: string; domains_used: number; max_domains: number }>(
    `SELECT id, alias, email, domains_used, max_domains
       FROM cf_keys
      WHERE is_active = 1 AND domains_used < max_domains`,
  )
  const totalActive = (all<{ n: number }>(`SELECT COUNT(*) AS n FROM cf_keys WHERE is_active = 1`)[0]?.n) ?? 0
  if (rows.length === 0) {
    if (totalActive === 0) return fail("No active CF keys in the pool. Add one in Settings.")
    return fail(
      `All ${totalActive} active CF keys are at capacity. ` +
      `Add a new key or raise max_domains on an existing one.`,
    )
  }
  const available = rows.reduce((s, r) => s + (r.max_domains - r.domains_used), 0)
  return ok(`${rows.length} key(s) with capacity, ${available} domain slot(s) free`)
}

export async function checkDoToken(): Promise<CheckResult> {
  const tok = (getSetting("do_api_token") || "").trim()
  if (!tok) return fail("DO_API_TOKEN not set in Settings")
  try {
    const r = await fetch("https://api.digitalocean.com/v2/account", {
      headers: { Authorization: `Bearer ${tok}` },
      signal: AbortSignal.timeout(15_000),
    })
    if (!r.ok) {
      return fail(`DO API rejected token (HTTP ${r.status})`, { body: (await r.text()).slice(0, 140) })
    }
    const j = (await r.json()) as { account?: { email?: string; status?: string; droplet_limit?: number } }
    const acc = j.account ?? {}
    return ok(
      `DO ok (${acc.email ?? "?"}, droplet_limit=${acc.droplet_limit ?? "?"})`,
      { email: acc.email, droplet_limit: acc.droplet_limit, status: acc.status },
    )
  } catch (e) {
    return fail(`DO API unreachable: ${(e as Error).name}: ${(e as Error).message}`)
  }
}

export async function checkSaAuth(): Promise<CheckResult> {
  const tok = (getSetting("serveravatar_api_key") || "").trim()
  const org = (getSetting("serveravatar_org_id") || "").trim()
  if (!tok) return fail("SERVERAVATAR_API_KEY not set in Settings")
  if (!org) return fail("SERVERAVATAR_ORG_ID not set in Settings")
  try {
    const r = await fetch(`https://api.serveravatar.com/organizations/${org}`, {
      headers: { Authorization: tok, Accept: "application/json" },
      signal: AbortSignal.timeout(15_000),
    })
    if (!r.ok) {
      return fail(`SA API rejected (HTTP ${r.status})`, { body: (await r.text()).slice(0, 140) })
    }
    return ok(`SA ok (org=${org})`)
  } catch (e) {
    return fail(`SA API unreachable: ${(e as Error).name}: ${(e as Error).message}`)
  }
}

export async function checkSpaceshipAuth(opts: { skipPurchase?: boolean } = {}): Promise<CheckResult> {
  if (opts.skipPurchase) return ok("Skipped (skip_purchase=True)")
  const apiKey = (getSetting("spaceship_api_key") || "").trim()
  const apiSecret = (getSetting("spaceship_api_secret") || "").trim()
  if (!apiKey || !apiSecret) return fail("SPACESHIP_API_KEY / _SECRET not both set")
  try {
    const r = await fetch("https://spaceship.dev/api/v1/domains?take=1", {
      headers: {
        "X-API-Key": apiKey,
        "X-API-Secret": apiSecret,
        Accept: "application/json",
      },
      signal: AbortSignal.timeout(15_000),
    })
    if (!r.ok) {
      return fail(`Spaceship API rejected (HTTP ${r.status})`, { body: (await r.text()).slice(0, 140) })
    }
    return ok("Spaceship ok")
  } catch (e) {
    return fail(`Spaceship unreachable: ${(e as Error).name}: ${(e as Error).message}`)
  }
}

export function checkLlmKey(): CheckResult {
  const provider = (getSetting("llm_provider") || "anthropic").trim().toLowerCase()
  const key = (getSetting(`llm_api_key_${provider}`) || "").trim() ||
              (getSetting("llm_api_key") || "").trim()
  if (!key) return fail(`LLM provider=${provider} but no API key configured`)
  return ok(`LLM provider=${provider} key configured (${key.length} chars)`)
}

export function checkServerCapacity(): CheckResult {
  const rows = all<{
    id: number; name: string; ip: string;
    sites_count: number; max_sites: number;
  }>(
    `SELECT s.id, s.name, s.ip,
            (SELECT COUNT(*) FROM domains d WHERE d.server_id = s.id) AS sites_count,
            s.max_sites
       FROM servers s
      WHERE s.status = 'ready'`,
  )
  const available = rows.filter((r) => (r.sites_count ?? 0) < (r.max_sites ?? 60))
  if (available.length > 0) {
    return ok(
      `${available.length} ready server(s) with capacity`,
      {
        servers: available.map((r) => ({
          name: r.name, ip: r.ip, sites: r.sites_count, max: r.max_sites,
        })),
      },
    )
  }
  if (rows.length > 0) {
    return fail(
      `All ${rows.length} ready server(s) are at max_sites. ` +
      `Pipeline will try to provision a new droplet.`,
    )
  }
  return fail(
    "No ready servers. Pipeline will try to provision a new droplet " +
    "(requires healthy DO token).",
  )
}

export function checkRootPassword(): CheckResult {
  const pw = (getSetting("server_root_password") || "").trim()
  if (!pw) {
    return fail(
      "server_root_password not set. New droplets won't be reachable via " +
      "password SSH for setup. Set in Settings.",
    )
  }
  return ok(`Root password set (${pw.length} chars)`)
}

/**
 * SSL custom-install fallback chain readiness.
 *
 * SA's REST /ssl endpoint returns HTTP 500 ("Something went wrong while
 * creating custom ssl certificate") for fresh apps — a documented SA-side
 * bug. The pipeline has a 3-tier API → UI → SSH fallback chain to handle
 * this. This check verifies the UI tier (the intended workaround) has
 * everything it needs:
 *
 *   1. patchright Chromium binary on disk
 *   2. sa_dashboard_email + sa_dashboard_password set in DB
 *   3. iproyal proxy configured (per ops policy: "patchright path uses
 *      iproyal; never use the bare machine IP")
 *
 * Returns `ok: true` regardless — the SSH tier is the unconditional
 * safety net, so the pipeline always completes. The message tells the
 * operator whether step 8 will succeed via the IDEAL path (UI, updates
 * SA dashboard tracker) or DEGRADE to SSH (cert is live but SA shows
 * "no SSL installed"). Non-blocking: missing patchright/proxy doesn't
 * stop a run, just downgrades the cosmetics.
 */
export function checkSslUiFallback(): CheckResult {
  // Patchright drops Chromium under <userCache>/ms-playwright/chromium*
  // — on Windows that's %LOCALAPPDATA%, on Linux/macOS ~/.cache.
  const candidates = [
    process.env.LOCALAPPDATA ? path.join(process.env.LOCALAPPDATA, "ms-playwright") : null,
    path.join(os.homedir(), ".cache", "ms-playwright"),
    path.join(os.homedir(), "Library", "Caches", "ms-playwright"),
  ].filter((p): p is string => !!p)
  let patchrightVersion = ""
  for (const dir of candidates) {
    if (!existsSync(dir)) continue
    const entries = readdirSync(dir).filter((d) => d.startsWith("chromium"))
    if (entries.length > 0) {
      patchrightVersion = entries[entries.length - 1]
      break
    }
  }

  const saEmail = (getSetting("sa_dashboard_email") || "").trim()
  const saPass = (getSetting("sa_dashboard_password") || "").trim()
  const proxyUrl = (getSetting("iproyal_proxy_url") || "").trim()
  const proxyServer = (getSetting("iproyal_proxy_server") || "").trim()

  const issues: string[] = []
  if (!patchrightVersion) {
    issues.push("patchright Chromium not installed (run: npx patchright install chromium)")
  }
  if (!saEmail || !saPass) {
    issues.push("sa_dashboard_email / sa_dashboard_password missing (UI fallback can't login)")
  }
  if (!proxyUrl && !proxyServer) {
    issues.push("iproyal_proxy_url not set (UI fallback would use bare IP — violates ops policy)")
  }

  if (issues.length === 0) {
    return ok(
      `SSL UI fallback ready (chromium=${patchrightVersion}, SA login + iproyal configured)`,
    )
  }
  // Always non-blocking — SSH tier is the unconditional safety net.
  return ok(
    `SSL UI fallback DEGRADED — pipeline will use SSH tier on fresh apps ` +
    `(cert live, but SA dashboard won't show installed=true). Issues: ${issues.join("; ")}`,
    { degraded: true, issues, patchrightVersion },
  )
}

// ---------------------------------------------------------------------------
// Aggregate
// ---------------------------------------------------------------------------

export interface PreflightReport {
  ok: boolean
  checks: Record<string, CheckResult>
}

export async function runAll(opts: { skipPurchase?: boolean } = {}): Promise<PreflightReport> {
  // Sequential — total budget ~10s on a healthy network. Could parallelize
  // but the order is informative for the dashboard (CF first since pipeline
  // step 2 needs it).
  const checks: Record<string, CheckResult> = {
    cf_pool:         checkCfPool(),
    do_token:        await checkDoToken(),
    sa_auth:         await checkSaAuth(),
    spaceship_auth:  await checkSpaceshipAuth({ skipPurchase: opts.skipPurchase }),
    llm_key:         checkLlmKey(),
    server_capacity: checkServerCapacity(),
    root_password:   checkRootPassword(),
    ssl_ui_fallback: checkSslUiFallback(),
  }
  return {
    ok: Object.values(checks).every((c) => c.ok),
    checks,
  }
}
